corehq/motech/value_source.py

```diff
@@ -49,9 +49,13 @@ def deserialize(self, external_value):
                       serializers.get((None, self.commcare_data_type)))
         return serializer(external_value) if serializer else external_value
 
-    def get_value(self, case_trigger_info):
+    def get_commcare_value(self, case_trigger_info):
         raise NotImplementedError()
 
+    def get_value(self, case_trigger_info):
+        value = self.get_commcare_value(case_trigger_info)
+        return self.serialize(value)
+
```

Comment:
nit: not sure if i understand this structure correctly, but if i do, there are now multiple methods `get_value`, `get_commcare_value` and then there is `serialize` to get the value. Is it possible to make a single method which should be called and others are hidden by making them private methods to keep the API a little simple.

Reply:
I agree with your point, and originally that single method was `get_value()`.

Then we found that when we imported values from external APIs (i.e. OpenMRS, Bahmni, or DHIS2) we wanted to call `deserialize()` on those values.

And now with this change we need to get a "CommCare" value before it's serialized so that we can manipulate it first, and then serialize it. ... An alternative could be to use `value_source.deserialize(value_source.get_value())` to deserialize the value that's just been serialized, but that feels backwards.

Maybe some docstrings would be in order. Something like,
```python
    def get_value(self, case_trigger_info):
        """
        This is the method you are looking for
        """
        # etc ...
```
:grinning:

Reply:
I like your idea ... *but* ... in different parts of the codebase the methods I'd want to make private (well, protected), `serialize()`, `deserialize()` and `get_commcare_value()` all get called from outside the object. `get_commcare_value()` only gets called from outside the object once, so maybe we could say "the exception proves the rule". :grin: ... I'll make that one protected.

Reply:
oh ho, okay